### PR TITLE
Update getSubmitToAccountID with category and tag approver

### PR DESCRIPTION
**Pull request by nkdengineer**
### Details

Update getSubmitToAccountID with category and tag approver
### Fixed Issues

$ https://github.com/Expensify/App/issues/51001
PROPOSAL: https://github.com/Expensify/App/issues/51001#issuecomment-2419241110

### Tests

Precondition:
- Create a WS and invite three users: employee, approver1, approver2
- Enable workflows and assign admin as an approver
- Add approver1 as approver of any category (`CAT_APPROVER`) in OD
- Add approver 2 as approver of any tag (`TAG_APPROVER`) in OD

1. [Employee] Go to WS chat and submit an expense with the category `CAT_APPROVER`
2. [Employee] Submit the expense and verify that the approver1 appears in the next step and the expense is submitted successfully
3. [Employee] Submit another expense with the tag `TAG_APPROVER`
4. [Employee] Submit the expense and verify that the approver2 appears in the next step and the expense is submitted successfully
5. [Employee] Submit two expenses, the first expense with the category `CAT_APPROVER` and the second one with the tag `TAG_APPROVER`
6. [Employee] Submit two expenses and verify that the approver1 appears in the next step and that the expense is submitted successfully
7. [Employee] Submit two expenses, the first expense with the tag `TAG_APPROVER` and the second one with the category `CAT_APPROVER`
8. [Employee] Submit two expenses and verify that the approver1 appears in the next step and that the expense is submitted successfully
- [x] Verify that no errors appear in the JS console

### Offline tests

Do the same step above and only need to verify that the approver display in the next step is correct.
### QA Steps

1. [Employee] Go to WS chat and submit an expense with the category `CAT_APPROVER`
2. [Employee] Submit the expense and verify that the approver1 appears in the next step and the expense is submitted successfully
3. [Employee] Submit another expense with the tag `TAG_APPROVER`
4. [Employee] Submit the expense and verify that the approver2 appears in the next step and the expense is submitted successfully
5. [Employee] Submit two expenses, the first expense with the category `CAT_APPROVER` and the second one with the tag `TAG_APPROVER`
6. [Employee] Submit two expenses and verify that the approver1 appears in the next step and that the expense is submitted successfully
7. [Employee] Submit two expenses, the first expense with the tag `TAG_APPROVER` and the second one with the category `CAT_APPROVER`
8. [Employee] Submit two expenses and verify that the approver1 appears in the next step and that the expense is submitted successfully
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/f9f23c1f-889d-48be-83a7-8c83f1866786

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/d6446d55-f126-4600-9ee0-dde43471aae1

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/21bf1a23-291d-478f-a13d-ec6a2d1b2388

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/2a5e1404-7d2e-4058-9670-fc4b897418f1

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/461f6d63-3e3d-4621-b854-95b8d25c7f42

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/cc70616e-af1d-420d-b3db-821e8c18aa4b

</details>

### CODE REVIEW

**Comment by suneox** [comment]
@Beamanator We have an issue where the approver for either a tag or a category cannot approve the expense. 

https://github.com/user-attachments/assets/86fbd90d-24eb-4b96-a805-52868764b00c

Have we got another rule from the backend regarding approvers for categories or tags? And is it included in the scope of this issue?

@nkdengineer I discovered an issue where the selected approver becomes unstable if there’s a value submission error

https://github.com/user-attachments/assets/f3b8841d-e65f-4c99-90d4-75f5956ec55b

**Comment by Beamanator** [comment]
Oof good to know! I'll look into this right meow

**Comment by Beamanator** [comment]
Ok so this is all a bit new at the moment (supporting category & tag approvers in NewDot) so we've def got some bugs to work out. Let's try to make each as clear & reproducible as possible so it's easy to resolve them 1 by 1 🙏 

1. I am noticing that when the submitter submits to a category/tag approver, that approver is not invited to the workspace chat, they only have access to the expense report. I believe that's a 🐛 
2. @suneox I see you somehow got an error `'Policy approve error'`, can you reproduce? I couldn't figure out how to reproduce that issue
3. And sorry but I'm a bit confused what exactly you're testing in your 2nd video & what the bug(s) are since it's a bit long 😅

**Comment by suneox** [comment]
> @suneox I see you somehow got an error 'Policy approve error', can you reproduce? I couldn't figure out how to reproduce that issue

We can reproduce the issue in this PR after an employee submits an expense has category X that requires approval from approver A on category X, then approver A approve the expense.

> And sorry but I'm a bit confused what exactly you're testing in your 2nd video & what the bug(s) are since it's a bit long

The issue happen from second 1:06 -> 1:11 (and 1:32 -> 1:39) approver change from `CP 51196` to `CP Approver 2`

**Comment by Beamanator** [comment]
> @Beamanator We can reproduce the issue in this PR after an employee submits an expense has category X that requires approval from approver A on category X, then approver A approve the expense.

@suneox what am I doing different here?

https://github.com/user-attachments/assets/6d25207c-245e-486e-8993-c819cfd178c5

**Comment by Beamanator** [comment]
> > And sorry but I'm a bit confused what exactly you're testing in your 2nd video & what the bug(s) are since it's a bit long
>
> The issue happen from second 1:06 -> 1:11 (and 1:32 -> 1:39) approver change from CP 51196 to CP Approver 2

I saw some things but again, it's only helpful if you have reproducible steps 🙏  from your video it looked like you clicked around kinda quickly & re-submitted even when you hit some errors instead of refreshing the page. It's useful to refresh to know if the error comes from the backend or the optimistic onyx data

**Comment by suneox** [comment]
> > The issue happen from second 1:06 -> 1:11 (and 1:32 -> 1:39) approver change from CP 51196 to CP Approver 2
> 
> I saw some things but again, it's only helpful if you have reproducible steps 🙏 from your video it looked like you clicked around kinda quickly & re-submitted even when you hit some errors instead of refreshing the page. It's useful to refresh to know if the error comes from the backend or the optimistic onyx data

I’m still trying to reproduce the issue but haven't found the exact steps to consistently reproduce it yet. So i will continue process with checklist and try it later

**Comment by Beamanator** [comment]
FYI I am seeing a bug locally where, when we are forwarding these category reports, initially the report action doesn't show up, but it will after a refresh -> I have a PR up to fix that

**Comment by suneox** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/9024a4ba-8117-4c0b-af0a-08ff67c847a3




</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/user-attachments/assets/6f85be20-ded4-4dc2-abf3-392a86eb235a



</details>


<details>
<summary>Android: Native</summary>


https://github.com/user-attachments/assets/01922afc-c9cf-4987-9f26-d4da4f573ef3



</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/user-attachments/assets/aa67c68c-a146-4f52-b6dd-c6ae6a0a1e29



</details>

<details>
<summary>iOS: Native</summary>


https://github.com/user-attachments/assets/5c1648eb-64aa-4cd0-9cb9-3325b0389dcf



</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/user-attachments/assets/2892d2d8-fabb-4268-af8d-346d77602884


</details>


Warning: `TRenderEngineProvider: Support for defaultProps` was raised on [slack](https://expensify.slack.com/archives/C049HHMV9SM/p1726072515490119?thread_ts=1726072125.690969&cid=C049HHMV9SM)

Warning: `Function components cannot be given refs` was raised on [slack](https://expensify.slack.com/archives/C049HHMV9SM/p1726072125690969)

Warning: `findDOMNode is deprecated in StrictMode` was raised on [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1723679073528139)

**Review by suneox — APPROVED** [review]
The change in the priority for selecting the approver LGTM. Regarding the issue I mentioned earlier in the conversation, it seems like it was a backend issue, and I'm now unable to reproduce it.

**Comment by Beamanator** [comment]
@nkdengineer would you mind pulling main? I have another backend PR to fix the issue in this flow with the correct / incorrect approver following a category/tag approver - if these work well together, i think we can get this merged soon!

**Comment by nkdengineer** [comment]
@Beamanator Just merged.

File: src/libs/PolicyUtils.ts
Comment (Beamanator) [inline src/libs/PolicyUtils.ts:523]: This needs to be updated, as we're no longer passing `employeeAccountID`

File: src/libs/PolicyUtils.ts
Comment (Beamanator) [inline src/libs/PolicyUtils.ts:525]: The only possibly annoying thing with this change is that now, we're getting the `submitTo` from the report owner ALWAYS - i believe we USED TO be able to pass any employee accountID & get their `submitTo` 🤔 

Technically that may not be a problem if this works everywhere & that's the only way we ever use this function 👍

File: src/libs/PolicyUtils.ts
Comment (Beamanator) [inline src/libs/PolicyUtils.ts:534]: Let's add a comment here explaining something like:

```suggestion
    // Before submitting to their `submitsTo` (in a policy on Advanced Approvals), submit to category/tag approvers.
    // Category approvers are prioritized, then tag approvers.
    for (let i = 0; i < allTransactions.length; i++) {
```

File: src/libs/ReportUtils.ts
Comment (Beamanator) [inline src/libs/ReportUtils.ts:8257]: Ooh this does make me happy

File: src/libs/ReportUtils.ts
Comment (Beamanator) [inline src/libs/ReportUtils.ts:8265]: Ooh we should probably be getting rule approvers here too & adding them to the approval chain like this:

`[...categoryApprovers, ...tagApprovers, ...baseApprovalChain]` (de-duped) 🤔

File: src/libs/PolicyUtils.ts
Comment (nkdengineer) [reply src/libs/PolicyUtils.ts:525]: @Beamanator Yes that's the only way we ever use this function. We only use this to get the `submitsTo` of the expense report.

File: src/libs/ReportUtils.ts
Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8287]: I think we don't need to do that here because `getSubmitToEmail` will return the `submitsTo` as `categoryApprover` or `tagApprover` if it is matched, and then we get this user's `forwardsTo`.

File: src/libs/PolicyUtils.ts
Comment (nkdengineer) [reply src/libs/PolicyUtils.ts]: I updated.

File: src/libs/PolicyUtils.ts
Comment (Beamanator) [reply src/libs/PolicyUtils.ts:525]: Ok then I think we're good to go here 👍

File: src/libs/ReportUtils.ts
Comment (Beamanator) [inline src/libs/ReportUtils.ts:3099]: Looks like there's a lint error above (line 1298) where another call to PolicyUtils.getSubmitToAccountID needs to get updated

File: src/libs/ReportUtils.ts
Comment (Beamanator) [reply src/libs/ReportUtils.ts:8287]: Hmm ok so `ReportUtils.getApprovalChain` gets called in 2 places so far:
1. IOU.approveMoneyRequest
    - Here we mainly check `isLastApprover` of the approval chain so it probably doesn't matter adding category/tag approvers to the front of the approval chain
    - We also call `getNextApproverAccountID` to optimistically figure out who the next manager should be
2. IOU.getNextApproverAccountID
    - This gets called from buildNextStep and other places (ex: ^)

I thinkkkkk we will see problems if we run some tests with multiple category/tag approvers, because we'll only try to calculate a category/tag approver when submitting, but not as the "next approver" - at least in the optimistic onyx data. (possibly even backend errors, but that's what i'm working on fixing)

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8287]: @nkdengineer can you test my theory ^ there? I can mention exact test steps if you need

**Comment by Beamanator** [comment]
@nkdengineer just checking in, did you end up testing out https://github.com/Expensify/App/pull/51196#discussion_r1816169977?

File: src/libs/ReportUtils.ts
Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8294]: > I can mention exact test steps if you need

@Beamanator Yes, please provide me the exact test steps.

Comment (garrettmknight) [reply src/libs/ReportUtils.ts:8294]: I think these are the steps to follow and confirm:
1. Create a workspace in NewDot
2. Enable delay submission
3. Invite submitter, Approver A, Approver B
4. Set to Advanced Approval in OldDot to workspace owner
5. Set category approvals for Category A going to Approver A + Category B as Approver B
6. As submitter in NewDot, submit 2 expenses to the WS chat, one with Category A, one with Category B
7. Submit the report
8. Confirm first approver is Approver A
9. As Approver A, approve the report
10. Confirm the second approver is Approver B
11. As Approver B, approve the report
12. Confirm the third approver is workspace owner
13. As WS owner, approve and confirm that the report is final approved.

Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8308]: @Beamanator While testing the steps below, in step 9 approver A can't approve the report. BE returns an error like this.


Screenshot 2024-10-31 at 23 36 49

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8308]: Thanks @garrettmknight - yes those look like great reproduction steps!

@nkdengineer - oof that's actually an old error that we should be getting rid of in https://github.com/Expensify/Web-Expensify/pull/44194, once that's deployed to staging (hopefully today) - I can try to ping here once we're ready for that deploy, but it probably won't happen till later (PST time)

**Comment by JmillsExpensify** [comment]
What's the latest on this PR? We're trying to wrap up support for this use case so that we can remove a related beta.

**Comment by nkdengineer** [comment]
We're waiting the backend fix [here](https://github.com/Expensify/App/pull/51196#discussion_r1827401999).

**Comment by Beamanator** [comment]
@nkdengineer the Web-E PR just went live, can you please retest?

**Comment by nkdengineer** [comment]
Sure.

File: src/libs/ReportUtils.ts
Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8383]: @Beamanator After approver A approves the money request, we're waiting for payment from admin. 

> I think we don't need to do that here because getSubmitToEmail will return the submitsTo as categoryApprover or tagApprover if it is matched, and then we get this user's forwardsTo

And it's what I said here.

https://github.com/user-attachments/assets/f8bef823-7ee9-4f89-937e-f62e7df277f3

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8383]: So it looks like after Approver A, we're expecting Approver B, right? I don't quite understand your comment though, sorry - did I lead you down a wrong path somewhere? 😅

Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8383]: @Beamanator The video below is what BE working on after the Approver A approves if this user doesn't have `forwardsTo` then the report is marked as approved.

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8383]: Hmmm ok i can test on my end - but can you just make sure this work for you, optimistically? (offline) If it works well here optimistically, we can get this moving & i can fix the backend later

Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8383]: So offline, we're expecting that after approver A approves the expense, the next one will be approver B right? I have a concern with this except if we have some approvers match with category/tag rules like A, B, C, and D. After the approvers A, and B approve the report, we don't have a way to detect the list of previous approves of the expense report.

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8383]: Yes exactly! And good call, that's an important consideration but I am ok with handling that in a separate PR - I just implemented this logic in the backend (not deployed yet) so i can help get this in the front end after this

Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8383]: Then the next step is I will handle `getApproveChain` logic here or in a separate PR.

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8383]: Hmm so the issue this PR is fixing is purely related to "Submitting" so IMO in this PR we don't need to update the getApprovalChain logic - that can be a fast follow-up

Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8383]: @Beamanator Got it, please check again. If it doesn't have any other problem. I think we're ready to merge this PR. Then we can create a follow-up later.

File: src/libs/PolicyUtils.ts
Comment (Beamanator) [inline src/libs/PolicyUtils.ts:533]: Just curious, is this called anywhere for `control` policies? If so, we shouldn't consider category / tag approvers, just the policy's `defaultApprover` (`approver` key)

File: src/libs/ReportUtils.ts
Comment (Beamanator) [reply src/libs/ReportUtils.ts:8383]: Will do - resolving this thread too

File: src/libs/PolicyUtils.ts
Comment (Beamanator) [reply src/libs/PolicyUtils.ts:533]: @nkdengineer my last question for this PR ^

**Comment by Beamanator** [comment]
So the plan will be:
1. Get this PR merged
2. Open next issue that accounts for a "better approval chain / next approver / receiver" code. In this issue we'll account for:
    1. Multiple category / tag approvers
    2. Managers who have been removed from the approval chain & who is next after they approve a report
    3. Combinations of all of the above

File: src/libs/PolicyUtils.ts
Comment (nkdengineer) [inline src/libs/PolicyUtils.ts:536]: @Beamanator We already have a variable `defaultApprover` here. It's `policy?.approver ?? policy?.owner ?? ''`

Comment (Beamanator) [reply src/libs/PolicyUtils.ts:536]: See comment below 🙏

File: src/libs/PolicyUtils.ts
Comment (nkdengineer) [reply src/libs/PolicyUtils.ts:563]: @Beamanator When I tested by adding the category/tag rules and didn't upgrade to advanced-approval, the category/tag rules is still applied.

Comment (Beamanator) [reply src/libs/PolicyUtils.ts:563]: That's great! But still isn't it better to return early via this check BEFORE looping through transactions? Not fixing a bug, but more efficient to return early

Comment (nkdengineer) [reply src/libs/PolicyUtils.ts:563]: @Beamanator I mean that the category/tag rule should still apply for non-advanced-approval policies. If we move this after this condition, the error appears when we submit the report of non-advanced-approval policies that has transaction match category/tag approver rule.

Comment (Beamanator) [reply src/libs/PolicyUtils.ts:563]: Oh hmmmm i assumed category/tag approvals were only supported w/ Advanced approval settings, do you know if that's incorrect @garrettmknight ?

Comment (Beamanator) [reply src/libs/PolicyUtils.ts:563]: Ok so I guessed wrong 😅  You're right @nkdengineer - category/tag rule approvers should still apply for non-advanced-approval Control policies

**Review by Beamanator — APPROVED** [review]

**Comment by Beamanator** [comment]
Let's get this merged - I'll be creating a new issue today for the approval-side of category/tag approvers & updating the approval chain / next receiver logic

**Pull request merged**
